ArjunaJTS/jts/classes/com/arjuna/ats/jts/extensions/AtomicTransaction.java

```diff
@@ -611,12 +611,14 @@ public boolean equals (Object obj)
 				AtomicTransaction tx = (AtomicTransaction) obj;
 				ControlWrapper txControl = tx._theAction;
 
-				if ((_theAction == null) && (txControl == null))
+				if ((_theAction == null) && (txControl == null)) {
 					return true;
-				else if (_theAction != null) {
+				}
+				else if ((_theAction != null) && (txControl==null)) {
```

Comment:
What is the problem that your addressing here?  Should this instead check for null _theAction, since the below _theAction.equals(txControl) check already handles the null txControl case:
else if (_theAction == null)
    return false;